File: cpp/leetcode/FrogJump.cpp
//Leetcode Problem No 403 Frog Jump
//Solution written by Xuqiang Fang on 22 June, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution{
public:
    bool canCross(vector<int>& stones) {
        if(stones[1] != 1){
            return false;
        }
        const int n = stones.size();
        unordered_map<int, unordered_set<int>> dp;
        for(const int& s : stones){
            dp.emplace(s, unordered_set<int>());
        }
        dp[1].emplace(1);
        for(int i=1; i<n-1; ++i){
            for(const int j : dp[stones[i]]){
                for(const int k : {j-1, j, j+1}){
                    if(k > 0 && dp.count(stones[i]+k)){
                        dp[stones[i]+k].emplace(k);
                    }
                }
            }
        }
        return !dp[stones[n-1]].empty();
    }
};

int main(){
    Solution s;
    vector<int> stones{0,1,3,5,6,8,12,17};
    cout << s.canCross(stones) << endl;
    stones = {0,1,2,3,4,8,9,11};
    cout << s.canCross(stones) << endl;
    return 0;
}

File: py/leetcode/FrogJump.py
'''
Leetcode problem No 403 Frog Jump 
Solution written by Xuqiang Fang on 22 June, 2018
'''

class Solution(object):
    def canCross(self, stones):
        """
        :type stones: List[int]
        :rtype: bool
        """
        if stones[1] != 1:
            return False
        n = len(stones)
        dp = {}
        for i in stones:
            dp[i] = set()
        dp[1].add(1)
        moves = (-1,0,1)
        for i in range(1,n-1):
            s = dp[stones[i]]
            if len(s) == 0:
                break
            for j in s:
                for x in moves:
                    key = stones[i] + j + x
                    if x+j > 0 and key in dp:
                        dp[key].add(x+j)

        return len(dp[stones[n-1]]) > 0
def main():
    s = Solution()
    stones = [0,1,3,5,6,8,12,17]
    print(s.canCross(stones))
    stones = [0,1,2,3,4,8,9,11]
    print(s.canCross(stones))
main()
